cms/server/contest/handlers/api.py

```diff
@@ -0,0 +1,147 @@
+            self.json({"error": "Login failed"}, 403)
+        elif login_data is not None:
+            cookie_name = self.contest.name + "_login"
+            self.json({"login_data": self.create_signed_value(
+                cookie_name, login_data).decode()})
+        else:
+            self.json({})
+
+    def check_xsrf_cookie(self):
+        pass
+
+
+class ApiTaskListHandler(ContestHandler):
+    """Handler to list all tasks and their statements.
+
+    """
+    @tornado_web.authenticated
+    @actual_phase_required(0, 3)
+    @multi_contest
+    def get(self):
+        contest = self.contest
+        tasks = dict()
+        for task in contest.tasks:
+            name = task.name
+            statements = [s for s in task.statements]
+            sub_format = task.submission_format
+            tasks[name] = {"statements": statements,
+                           "submission_format": sub_format}
+        self.json(tasks)
+
+
+class ApiSubmitHandler(ContestHandler):
+    """Handles the received submissions.
+
+    """
+    @tornado_web.authenticated
+    @actual_phase_required(0, 3)
+    @multi_contest
+    def post(self, task_name: str):
+        task = self.get_task(task_name)
+        if task is None:
+            self.json({"error": "Not found"}, 404)
+            return
+
+        # Only set the official bit when the user can compete and we are not in
+        # analysis mode.
+        official = self.r_params["actual_phase"] == 0
+
+        try:
+            submission = accept_submission(
+                self.sql_session, self.service.file_cacher, self.current_user,
+                task, self.timestamp, self.request.files,
+                self.get_argument("language", None), official)
+            self.sql_session.commit()
+        except UnacceptableSubmission as e:
+            logger.info("API submission rejected: `%s' - `%s'",
+                        e.subject, e.formatted_text)
+            self.json({"error": e.subject, "details": e.formatted_text}, 400)
+        else:
+            logger.info(
+                f'API submission accepted: Submission ID {submission.id}')
+            self.service.evaluation_service.new_submission(
+                submission_id=submission.id)
+            self.json({'id': submission.opaque_id})
+
+
+class ApiSubmissionListHandler(ContestHandler):
+    """Retrieves the list of submissions on a task.
+
+    """
+    @tornado_web.authenticated
+    @actual_phase_required(0, 3)
+    @multi_contest
+    def get(self, task_name: str):
+        task = self.get_task(task_name)
+        if task is None:
+            self.json({"error": "Not found"}, 404)
+            return
+        submissions: list[Submission] = (
+            self.sql_session.query(Submission)
+            .filter(Submission.participation == self.current_user)
+            .filter(Submission.task == task)
+            .all()
+        )
+        self.json({'list': [s.opaque_id for s in submissions]})
```

Comment:
I would return an array of objects, even though the objects will contain just the ID. It would make it easy to add more attributes of submissions in the future.